src/wallet/rpcwallet.cpp

```diff
@@ -326,36 +326,51 @@ static UniValue setlabel(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
-
-static CTransactionRef SendMoney(interfaces::Chain::Lock& locked_chain, CWallet * const pwallet, const CTxDestination &address, CAmount nValue, bool fSubtractFeeFromAmount, const CCoinControl& coin_control, mapValue_t mapValue)
+UniValue SendMoney(interfaces::Chain::Lock& locked_chain, CWallet * const pwallet, CCoinControl coin_control, UniValue& address_amounts, mapValue_t map_value, UniValue& subtract_fee_outputs)
 {
-    CAmount curBalance = pwallet->GetBalance(0, coin_control.m_avoid_address_reuse).m_mine_trusted;
+    std::set<CTxDestination> destinations;
+    std::vector<CRecipient> recipients;
+    for (const std::string& address: address_amounts.getKeys()) {
+        CTxDestination dest = DecodeDestination(address);
+        if (!IsValidDestination(dest)) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, std::string("Invalid Bitcoin address: ") + address);
+        }
+
+        if (destinations.count(dest)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("Invalid parameter, duplicated address: ") + address);
+        }
+        destinations.insert(dest);
+
+        CScript script_pub_key = GetScriptForDestination(dest);
+        CAmount amount = AmountFromValue(address_amounts[address]);
```

Comment:
This (order N, so in aggregate, quadratic) lookup would be necessary if the loop would iterate over key,value pairs in `address_amounts`. But I don't think univalue offers this functionality?

Reply:
There's `getKeys()` and `getValues()` but not both. A future refactor could convert from `UniValue` to `std::map<CTxDestination, CAmount>`, but you have to loop over the `UniValue` at least once. So if it's really a performance bottleneck for big transactions, someone would need to patch UniValue to add e.g. `bool getStringDictionary(std::map<std::string,std::string>& dict)`

Reply:
Can also do

```c++
int i = 0;
for (const std::string& address: address_amounts.getKeys()) {
    CAmount amount = AmountFromValue(address_amounts[i++]);
    ...
```

Reply:
Ah, I didn't notice this operator: `const UniValue& operator[](size_t index) const;`